### PR TITLE
Introduce safetensors

### PR DESCRIPTION
**Pull request by vince62s**
In order to make no break changes here is the approach:

When converting for instance llama13B which has 2 shards (consolidated.00.pth, consolidated.01.pth) we will create:
- 1 pytorch file with keys 'opt', 'vocab', 'optim' (same content as the regular .pt file) name being "mymodel.pt"
- N safetensors files containing the tensors of 'model' and 'generator', in N files named: mymodel.X.safetensors (X from 0 to 9 so far)

When finetuning, each shard will be loaded in the model one by one, hence preventing from saturating the Ram.
In the config the train_from needs to be the "mymodel.pt"

When training from scratch, a new option is introduced "save_format" which would enable to save the keys 'model' and 'generator' not in the .pt file but in a .safetensors file.

@francoishernandez do you have any comment on this ?

Support `use_embeddings_from_checkpoint()` for vocab update with safetensors left for later.
Add optional sharding at save time in safetensors format left for later(most use cases will be LoRA diff)

### CODE REVIEW

File: onmt/model_builder.py [hunk 1]
```diff
@@ -375,9 +383,26 @@ def build_model(model_opt, opt, vocabs, checkpoint):
...
+                precision=precision,
+                device=device,
+                strict=strict,
+            )
+        else:
+            # weights are not in the .pt checkpoint but stored in the safetensors file
+            base_name = (
+                opt.train_from[:-3] if opt.train_from[-3:] == ".pt" else opt.train_from
+            )
+            model_path = base_name
```
Comment (francoishernandez) [inline onmt/model_builder.py:399]: Specific reason for adding this intermediary name?

File: onmt/model_builder.py [hunk 2]
```diff
@@ -106,12 +106,20 @@ def load_test_model(opt, model_path=None):
...
+    if "model" in checkpoint.keys():
+        # weights are in the .pt file
+        model.load_state_dict(
+            checkpoint, precision=torch.float32, device=torch.device("cpu"), strict=True
+        )
+    else:
+        # weights are not in the .pt checkpoint but stored in the safetensors file
+        base_name = model_path[:-3] if model_path[-3:] == ".pt" else model_path
+        model_path = base_name + ".safetensors"
+        model.load_sf_state_dict(
```
Comment (francoishernandez) [inline onmt/model_builder.py:118]: typo `load_safe_state_dict`?

File: onmt/model_builder.py [hunk 3]
```diff
@@ -106,12 +106,20 @@ def load_test_model(opt, model_path=None):
...
+            model_path, precision=torch.float32, device=torch.device("cpu"), strict=True
```
Comment (francoishernandez) [inline onmt/model_builder.py:119]: Will it always be float32 precision here?

File: onmt/models/model.py [hunk 4]
```diff
@@ -100,6 +101,64 @@ def load_state_dict(
...
+        try:
+            import safetensors
+        except ImportError:
+            raise ImportError("run: pip install safetensors, to use safetensors")
+        keyfound = {}
+        shards = glob.glob(model_path + ".*.safetensors")
+        f = []
+        keys_shard = {}
+        for i, shard in enumerate(shards):
+            f.append(safetensors.safe_open(shard, framework="pt", device="cpu"))
```
Comment (francoishernandez) [inline onmt/models/model.py:134]: device -> use arg passed in the function?

**Review by francoishernandez — COMMENTED** [review]
Small comments.
I might have missed it but I don't see where we specify the split (from 0 to 9), is it hardcoded/default in safetensors?

File: onmt/model_builder.py (hunk 1 above)
Comment (vince62s) [reply onmt/model_builder.py:399]: no

File: onmt/models/model.py (hunk 4 above)
Comment (vince62s) [reply onmt/models/model.py:134]: no, the way it works is:
each module (the whole model) is built empty on cpu. we load the data of each module on cpu, and as each module is loaded with the tensors, it is moved to gpu (this is exactly at this step that data is quantized).
see the module.to(device) juste after.

File: requirements.opt.txt [hunk 5]
```diff
@@ -5,3 +5,4 @@ subword-nmt>=0.3.7
 rapidfuzz
 scipy
 bitsandbytes>=0.39.0
+safetensors
```
[the remark this answers is not in the document]
Comment (vince62s) [reply requirements.opt.txt:8]: I suggest we'll leave it open and when it breaks it will trigger a signal.

File: onmt/modules/lora.py [hunk 6]
```diff
@@ -195,13 +195,18 @@ def train(self, mode: bool = True):
...
 
             def forward(self, x: torch.Tensor):
-                result = self.maybe_ckpt(super().forward, x)
                 if self.r > 0 and not self.merged:
-                    result += (
-                        self.lora_dropout(x)
-                        @ self.lora_A.transpose(0, 1)
-                        @ self.lora_B.transpose(0, 1)
-                    ) * self.scaling
+                    result = (
```
Comment (vince62s) [inline onmt/modules/lora.py:199]: in some instance I am getting an inplace backward error when using results += ...

**Pull request merged**